### PR TITLE
Adding OFT support

### PR DESCRIPTION
**Pull request by antis0007**
A WIP adaptation of the OFT implementation from the Kohya repo

## Description

Adds OFT support by reimplementing networks/oft.py from the kohya-ss sd-scripts repo.

## Notes
WIP at the moment, and may not be implemented entirely correctly. Working with @AI-Casanova on this through discord.
This OFT implementation only supports SDXL (based on the one from sd-scripts)

## Environment and Testing

Developed on Windows 10, tested on Windows 10, dev branch

### CODE REVIEW

File: extensions-builtin/Lora/network_oft.py [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+        """
+        dim -> num blocks
+        alpha -> constraint
+
+        alpha is equal to eps-deviation: eps
+        (only with the constrained variant COFT)
+        """
+        self.weights = weights.w.get("oft_blocks").to(device=devices.device)
+        self.net = net
+        self.alpha = self.multiplier()
```
Comment (AI-Casanova) [inline extensions-builtin/Lora/network_oft.py:49]: Alpha does not equal a multiplier here. Should be able to set it to 1 and forget about it, as its only a training constraint and not involved in inference.

**Comment by AI-Casanova** [comment]
I see what you mean about the weirdness of the matmul shapes

Gotta hit the sack, will pick back up when I can.

**Comment by vladmandic** [comment]
looking forward to this!

**Comment by AI-Casanova** [comment]
I think I'm getting somewhere, but at the end of the day I may need a bit of help with the matrix math.

Will post my findings this evening

**Comment by AI-Casanova** [comment]
@antis0007 Please test my newest commit. Massive refactor, but should give very similar results.

Correct scaling is still an open question.

**Comment by vladmandic** [comment]
can you either rebase with dev or create a new pr based on your branch? there are many ghost changes here purely because of differences between master and dev at the time of creation of the pr.

otherwise, changes look really clean, no issues from my side.

**Comment by antis0007** [comment]
Not sure if that rebase worked, but let me know. I can also just take the files we changed, open a new fork from dev, then open a new PR to the dev branch if you'd like that.

**Comment by vladmandic** [comment]
> Not sure if that rebase worked, but let me know. I can also just take the files we changed, open a new fork from dev, then open a new PR to the dev branch if you'd like that.

its looking good.

**Pull request merged**